### ISSUE TITLE
Flushables should always be flushed

### ISSUE THREAD

**Issue by bjourne** [factor/factor#1568]
Let us define three words:
```factor
IN: scratchpad : foo ( -- ) 3 f <array> drop ;
IN: scratchpad : bar ( -- ) "test" f <array> drop ;
IN: scratchpad : baz ( -- ) -3 f <array> drop ;
IN: scratchpad { foo bar baz } [ optimized. ] each
[ ]
[ "test" f <array> drop ]
[ ]
```
As you can see, the optimizer doesn't flush away the middle word. It probably should flush all of them. That implies that if you type `-3 f <array> drop` on the listener prompt something different happens than if you type `: test ( -- ) -3 f <array> drop ; test` but I don't think that is a problem.

**Comment by bjourne** [comment]
Looks like it can't be done. With my fix this word in compiler.tests:
```factor
: aggressive-flush-regression ( a -- b ) 
    f over [ <array> drop ] dip 1 + ;
```
is optimized to:
```factor
[
    >R R> 1 over tag 0 eq?
    [ fixnum+ ] [ fixnum>bignum bignum+ ] if
]
```
Which is wrong. It probably happens because type propagation is run before dead code removal.

**Comment by jonenst** [comment]
So I'm guessing you're saying it's wrong because the compiled result is unsafe (it crashes badly for inputs that are not fixnums are bignums), but the code is safe.

I think it's also wrong to not throw the Invalid array size exception for negative and huge (> max-array-capacity) numbers. That is, I disagree with this:
> That implies that if you type -3 f <array> drop on the listener prompt something different happens than if you type : test ( -- ) -3 f <array> drop ; test but I don't think that is a problem.
makes it harder to reason about the program because removing the ```drop``` 

That's what is currently implemented (by you I think?) now (d56b0cbf23b), which I think is better than the original flush-all proposal:
```factor
IN: scratchpad : foo ( -- ) 3 f <array> drop ;
IN: scratchpad : bar ( -- ) "test" f <array> drop ;
IN: scratchpad : baz ( -- ) -3 f <array> drop ;
IN: scratchpad { foo bar baz } [ optimized. ] each
[ ]
[ "test" f <array> drop ]
[ -3 f <array> drop ]
```

So I think we can close this ?

**Comment by mrjbq7** [comment]
Good idea, closing. Thanks @bjourne and @jonenst.
